tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/ChooseOrGuessRule.scala

```diff
@@ -0,0 +1,87 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt._
+import at.forsyte.apalache.tla.bmcmt.rules.aux.{CherryPick, OracleHelper}
+import at.forsyte.apalache.tla.lir.{OperEx, SetT1, TlaType1}
+import at.forsyte.apalache.tla.lir.TypedPredefs._
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.oper.{ApalacheOper, TlaOper}
+
+/**
+ * <p>Rewriting rule for CHOOSE and Apalache!Guess. We implement a non-deterministic choice. It is not hard to add the
+ * requirement of determinism, but that will most likely affect efficiency.</p>
+ *
+ * <p>We will make CHOOSE deterministic in the issue #841 is solved.</p>
+ *
+ * @author
+ *   Igor Konnov
+ */
+class ChooseOrGuessRule(rewriter: SymbStateRewriter) extends RewritingRule {
```

Comment:
This would be easier to review if the refactoring, introduction of `Guess`, and renaming of the class were in separate commits. My brain's `diff` implementation is slow...